zerver/lib/message.py

```diff
@@ -1697,3 +1697,109 @@ def update_to_dict_cache(
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.UNMUTED
+
+    return None
+
+
+def visibility_policy_for_send_message(
+    sender: UserProfile,
+    message: Message,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a message
+    is sent to a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    # We prioritize 'FOLLOW' over 'UNMUTE'.
+    # We need to carefully handle the following two cases:
+    #
+    # 1. When a user starts a topic
+    #    note: starting a topic is INITIATION as well as PARTICIPATION
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |           ON_INITIATION              |     FOLLOWED
+    #         ON_INITIATION                |          ON_PARTICIPATION            |     FOLLOWED
+    #         ON_INITIATION                |            ON_INITIATION             |     FOLLOWED
+    #
+    # 2. When both the policies have the same value 'ON_PARTICIPATION'.
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |          ON_PARTICIPATION            |     FOLLOWED
+    visibility_policy = None
+
+    if sender.is_bot:
+        return visibility_policy
+
+    visibility_policy = visibility_policy_for_participation(sender)
+
+    if visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+        return visibility_policy
+
+    accessible_messages_in_topic = UserMessage.objects.filter(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    )
+
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
+        and accessible_messages_in_topic.count() == 1
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
+        and accessible_messages_in_topic.count() == 1
+    ):
+        return UserTopic.VisibilityPolicy.UNMUTED
+
+    return visibility_policy
+
+
+def should_change_visibility_policy(
+    new_visibility_policy: int,
+    sender: UserProfile,
+    stream_id: int,
+    topic_name: str,
+) -> bool:
+    user_topic = UserTopic.objects.filter(
+        user_profile=sender, stream_id=stream_id, topic_name__iexact=topic_name
+    )
+    if not user_topic:
+        return True
+    current_visibility_policy = user_topic[0].visibility_policy
```

Comment:
I would usually write this using `.get` and just catching the `UserTopic.DoesNotExist` exception to do the `return True` case.

Reply:
Done.